src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,441 @@
+                                    .map_err(|err| format!("Error overwriting symlink for local execution in workdir {:?}: {:?}", &workdir_for_server2, err))
+                            });
+                        futures::future::result(res)
+                    } else {
+                        debug!("JDK home for Nailgun already exists in {:?}. Using that one.", &workdir_for_server4);
+                        futures::future::ok(())
+                    }
+                }
+        })
+        .inspect(move |_| debug!("Materialized directory {:?} before connecting to nailgun server.", &workdir_for_server3))
+        .to_boxed()
+    }
+
+    fn get_python_distribution(&self) -> String {
+        format!("{}", self.python_distribution_absolute_path.display())
+    }
+}
+
+impl super::CommandRunner for NailgunCommandRunner {
+    fn run(
+        &self,
+        req: MultiPlatformExecuteProcessRequest,
+        context: Context) -> BoxFuture<FallibleExecuteProcessResult, String> {
+
+        let nailgun_pool = self.nailgun_pool.clone();
+        let inner = self.inner.clone();
+        let python_distribution = self.get_python_distribution();
+
+        let client_req = self.extract_compatible_request(&req).unwrap();
+
+        if !client_req.is_nailgunnable {
+            trace!("The request is not nailgunnable! Short-circuiting to regular process execution");
+            return self.inner.run(req, context)
+        }
+        debug!("Running request under nailgun:\n {:#?}", &client_req);
+
+        // Separate argument lists, to form distinct EPRs for (1) starting the nailgun server and (2) running the client in it.
+        let ParsedJVMCommandLines { nailgun_args, client_args, client_main_class } = ParsedJVMCommandLines::parse_command_lines(&client_req.argv).expect("Error parsing command lines!!");
+        let nailgun_req = construct_nailgun_server_request(nailgun_args, client_req.jdk_home.clone(), client_req.target_platform);
+        trace!("Extracted nailgun request:\n {:#?}", &nailgun_req);
+
+        let nailgun_req_digest = crate::digest(MultiPlatformExecuteProcessRequest::from(nailgun_req.clone()), &self.metadata);
+
+        let nailgun_name = NailgunCommandRunner::calculate_nailgun_name(&client_main_class);
+        let nailgun_name2 = nailgun_name.clone();
+        let nailgun_name3 = nailgun_name.clone();
+
+        let workdir_for_this_nailgun = try_future!(self.get_nailguns_workdir(&nailgun_name));
+        let workdir_for_this_nailgun1 = workdir_for_this_nailgun.clone();
+
+        self.materialize_workdir_for_server(workdir_for_this_nailgun.clone(), nailgun_req.jdk_home.clone(), client_req.input_files, context.workunit_store.clone())
+            .and_then(move |_metadata| {
+                // Connect to a running nailgun.
+                nailgun_pool.connect(nailgun_name.clone(), nailgun_req, &workdir_for_this_nailgun1, nailgun_req_digest)
```

Comment:
Should this be spawned on `task_executor::Executor::spawn_on_io_pool`? (If this is going to be slow and blocking, the answer is probably yes)